prov/rxm/src/rxm.h

```diff
@@ -292,12 +295,29 @@ struct rxm_tx_entry {
 		struct rxm_rma_buf *rma_buf;
 	};
 
-	/* Used for large messages and RMA */
-	struct fid_mr *mr[RXM_IOV_LIMIT];
-	struct rxm_rx_buf *rx_buf;
+	union {
+		/* Used for large messages and RMA */
+		struct {
+			struct fid_mr *mr[RXM_IOV_LIMIT];
+			struct rxm_rx_buf *rx_buf;
+		};
+		/* Used for SAR protocol */
+		struct {
+			size_t seg_cnt;
+			uint64_t msg_id;
+			struct dlist_entry in_flight_tx_buf_list;
+			struct rxm_iov rxm_iov;
+			uint64_t iov_offset;
+		};
+	};
 };
 DECLARE_FREESTACK(struct rxm_tx_entry, rxm_txe_fs);
 
+struct rxm_sar_key {
+	uint64_t msg_id;
+	struct fid_ep *msg_ep;
+};
```

Comment:
Yes, it's possible. But the rbtree/hash is used to avoid lookup of the RX entry that requires O(n) like searching entry in list.

Let's assume a peer posts N fi_recv and a remote peer does fi_send N times (each size fits into SAR protocol). The sender splits buffer into several portions and sends them to the peer.

In that time, peer receives first portion of each message and does for them tag:addr tuple matching and then assigns an approach RX entry. Then the matched RX entry is used to received remaining portions to reassembly messages.

We need to maintain a list (or hash like Arun proposed)  with already matched RX entries to eliminate spending time for each time tag:addr-matching for each received segment.

Reply:
Hmm, I was wrong in my previous comment.

Yes, we can maintain a list of RX entry. In simple cases this list should contain single RX entry, because the sender does a bunch of fi_send for all segments.
In case if there is no buffers (or core provider reports EAGAIN), RxM enqueues the operation and waiting a completions and does fi_send()s again and again. So, in such case we can receive segments from different messages and we should lookup an appropriate RX entry from the list.

The list is maintained per rxm_conn (i.e. per MSG EP).

Reply:
So, we can just use msg_id to lookup an appropriate RX entry in the list